File: qabelbox/src/main/java/de/qabel/qabelbox/fragments/BaseFragment.java
package de.qabel.qabelbox.fragments;

import android.app.Fragment;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.ActionBar;
import android.view.View;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import de.qabel.qabelbox.R;
import de.qabel.qabelbox.activities.MainActivity;
import de.qabel.qabelbox.dagger.HasComponent;
import de.qabel.qabelbox.listeners.IdleCallback;

/**
 * Base Fragment
 * Created by danny on 08.01.2016.
 */
public abstract class BaseFragment extends Fragment {

    protected static Executor serialExecutor = Executors.newSingleThreadExecutor();
    protected ActionBar actionBar;

    protected MainActivity mActivity;

    @Nullable
    protected IdleCallback idle;

    public void setIdleCallback(@Nullable IdleCallback idle) {
        this.idle = idle;
    }

    public void busy() {
        if (idle != null) {
            idle.busy();
        }
    }

    public void idle() {
        if (idle != null) {
            idle.idle();
        }
    }

    @SuppressWarnings("unchecked")
    protected <C> C getComponent(Class<C> componentType) {
        return componentType.cast(((HasComponent<C>)getActivity()).getComponent());
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mActivity = (MainActivity) getActivity();
        actionBar = mActivity.getSupportActionBar();
    }

    /**
     * @return title for fragment
     */
    public String getTitle() {
        return getString(R.string.app_name);
    }

    /**
     * @return true if floating action button used
     */
    public boolean isFabNeeded() {

        return false;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mActivity = null;
        actionBar = null;
    }

    @Override
    public void onResume() {

        super.onResume();
        if (actionBar != null) {
            actionBar.setTitle(getTitle());
        }
        if (isFabNeeded()) {
            mActivity.fab.show();
        } else {
            mActivity.fab.hide();
        }
    }

    /**
     * set own back listener in actionbar
     */
    protected void setActionBarBackListener() {

        mActivity.toggle.setToolbarNavigationClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                mActivity.onBackPressed();
            }
        });
    }

    /**
     * set own back listener in actionbar
     */
    protected void setActionBarBackListener(final View.OnClickListener listener) {

        mActivity.toggle.setToolbarNavigationClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                if (listener != null) {
                    listener.onClick(v);
                }
                mActivity.onBackPressed();
            }
        });
    }

    /**
     * @return true if fragment handle back button. otherwise return false to display sideMenu icon
     */
    public boolean supportBackButton() {

        return false;
    }

    /**
     * @return true if fragment handle back button. otherwise return false to display sideMenu icon
     */
    public boolean supportSubtitle() {

        return false;
    }

    /**
     * handle hardware back button
     */
    public void onBackPressed() {

    }

    public void updateSubtitle() {

    }

    public boolean handleFABAction() {

        return false;
    }
}

File: qabelbox/src/main/java/de/qabel/qabelbox/fragments/BaseFragment.kt
package de.qabel.qabelbox.fragments

import android.app.Fragment
import android.os.Bundle
import android.support.v7.app.ActionBar
import de.qabel.qabelbox.R
import de.qabel.qabelbox.activities.MainActivity
import de.qabel.qabelbox.dagger.HasComponent
import de.qabel.qabelbox.listeners.IdleCallback
import de.qabel.qabelbox.listeners.IntentListener
import de.qabel.qabelbox.listeners.toIntentFilter
import de.qabel.qabelbox.ui.QblView
import org.jetbrains.anko.ctx
import org.jetbrains.anko.longToast
import org.jetbrains.anko.onUiThread


abstract class BaseFragment(protected val mainFragment: Boolean = false,
                            protected val showOptionsMenu: Boolean = false,
                            protected val showFABButton: Boolean = false) : Fragment(), QblView {

    protected var actionBar: ActionBar? = null

    protected var mActivity: MainActivity? = null

    protected var idle: IdleCallback? = null

    protected val intentListeners = emptyList<IntentListener>()

    fun setIdleCallback(idle: IdleCallback?) {
        this.idle = idle
    }

    fun busy() = idle?.let { it.busy() }

    fun idle() = idle?.let { it.idle() }

    @Suppress("UNCHECKED_CAST")
    protected fun <C> getComponent(componentType: Class<C>): C {
        return componentType.cast((activity as HasComponent<C>).component)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        mActivity = activity as MainActivity
        actionBar = mActivity?.supportActionBar
    }

    /**
     * @return title for fragment
     */
    open val title: String
        get() = getString(R.string.app_name)

    open val subtitle: String? = null

    override fun onDetach() {
        super.onDetach()
        mActivity = null
        actionBar = null
    }

    override fun onResume() {
        super.onResume()
        refreshToolbarTitle()
        if (mainFragment) {
            configureAsMainFragment()
        } else {
            configureAsSubFragment()
        }

        if (showFABButton) {
            mActivity?.fab?.show()
        } else {
            mActivity?.fab?.hide()
        }
        setHasOptionsMenu(showOptionsMenu)

        intentListeners.forEach {
            ctx.registerReceiver(it.receiver, it.toIntentFilter())
        }
    }

    override fun onPause() {
        intentListeners.forEach {
            ctx.unregisterReceiver(it.receiver)
        }
        super.onPause()
    }

    protected fun refreshToolbarTitle() {
        actionBar?.apply {
            this.title = this@BaseFragment.title
            this.subtitle = this@BaseFragment.subtitle
        }
    }

    protected fun configureAsSubFragment() {
        mActivity?.toggle?.isDrawerIndicatorEnabled = false
        actionBar?.setDisplayHomeAsUpEnabled(true)
        setActionBarBackListener()
    }

    protected fun configureAsMainFragment() {
        actionBar?.setDisplayHomeAsUpEnabled(false)
        mActivity?.toggle?.isDrawerIndicatorEnabled = true
    }

    protected fun setActionBarBackListener() {
        mActivity?.toggle?.setToolbarNavigationClickListener { mActivity?.onBackPressed() }
    }

    /**
     * handle hardware back button
     *
     * @return true if the event was handled
     */
    open fun onBackPressed(): Boolean {
        return false
    }

    open fun handleFABAction(): Boolean {
        return false
    }

    override fun showDefaultError(throwable: Throwable) {
        onUiThread {
            longToast(throwable.message ?: "Error")
            error("Error", throwable)
        }
    }
}
